projects/core/src/cms/page/routing/page-link.factory.spec.ts

```diff
@@ -0,0 +1,202 @@
+      });
+
+      it(`should not add https`, () => {
+        expect(
+          service.resolveCanonicalUrl(
+            { forceHttps: false },
+            'http://test.com/xyz'
+          )
+        ).toEqual('http://www.test.com/xyz/');
+      });
+
+      it(`should not replace www. in redirects`, () => {
+        const canonical = service.resolveCanonicalUrl(
+          { forceHttps: true, removeQueryParams: false, forceWww: false },
+          'http://test.com/xyz?redirect=http://redirect.com'
+        );
+        expect(canonical).toContain('https://test.com');
+        expect(canonical).toContain('http://redirect.com');
+      });
+    });
+
+    describe('forceWww', () => {
+      it(`should add www. subdomain`, () => {
+        expect(
+          service.resolveCanonicalUrl({ forceWww: true }, 'http://test.com/xyz')
+        ).toContain('www.');
+      });
+
+      it(`should not add www. subdomain`, () => {
+        expect(
+          service.resolveCanonicalUrl(
+            { forceWww: false },
+            'http://test.com/xyz'
+          )
+        ).not.toContain('www.');
+      });
+
+      it(`should not add www. subdomain to redirects`, () => {
+        const canonical = service.resolveCanonicalUrl(
+          { forceWww: true, removeQueryParams: false, forceHttps: false },
+          'http://test.com/xyz?redirect=http://redirect.com'
+        );
+        expect(canonical).toContain('http://www.test.com');
+        expect(canonical).toContain('http://redirect.com');
+      });
+    });
+
+    describe('forceTrailingSlash', () => {
+      it(`should add trailing slash`, () => {
+        const canonical = service.resolveCanonicalUrl(
+          {
+            forceTrailingSlash: true,
+          },
+          'https://www.storefront.com?foo=bar'
+        );
+        expect(canonical).toEqual('https://www.storefront.com/');
+      });
+
+      it(`should not add trailing slash`, () => {
+        const canonical = service.resolveCanonicalUrl(
+          {
+            forceTrailingSlash: true,
+            removeQueryParams: false,
+          },
+          'https://www.storefront.com?foo=bar'
+        );
+        expect(canonical).toEqual('https://www.storefront.com?foo=bar');
+      });
+
+      it(`should add trailing slash when params are removed`, () => {
+        const canonical = service.resolveCanonicalUrl(
+          {
+            forceTrailingSlash: true,
+            removeQueryParams: true,
+          },
+          'https://www.storefront.com?foo=bar'
+        );
+        expect(canonical).toEqual('https://www.storefront.com/');
+      });
+
+      it(`should add trailing slash`, () => {
+        const canonical = service.resolveCanonicalUrl(
+          {
+            forceTrailingSlash: true,
+          },
+          'https://www.storefront.com/page'
+        );
+        expect(canonical).toEqual('https://www.storefront.com/page/');
+      });
+
+      it(`should add trailing slash`, () => {
```

Comment:
[nitpick] The test describes the case when we should not add trailing slash
```suggestion
      it(`should not add trailing slash`, () => {
```